### ISSUE TITLE
HashGraph allows duplicate edges when it shouldn't.

### ISSUE THREAD

**Issue by adamnovak** [vgteam/libbdsg#39]
`create_edge` is supposed to "ignore" existing edges, which to me means that if you add the same edge twice it exists only once.

https://github.com/vgteam/libhandlegraph/blob/e2c45aae9a19a14176e3dc412b6ebc038385a31b/src/include/handlegraph/mutable_handle_graph.hpp#L30-L31

But HashGraph seems to accept any number of duplicates of an edge. There's no pre-existence check:

https://github.com/vgteam/libbdsg/blob/7b913da50dc70bf2b4b301e1226a6586e27892bc/src/hash_graph.cpp#L37-L41

HashGraph should deduplicate edges.

**Comment by adamnovak** [comment]
@jeizenga Can you fix this? Or should I try to?

**Comment by jeizenga** [comment]
Are we sure we want to provide this guarantee? If we do that, we either need to add a reasonably heavy-duty data structure for edge lookup or else accept O(N) edge addition, which I think could blow up on some of the multiple sequence alignment-based graphs. What's the need for this check, as opposed to leaving it UB?

**Comment by adamnovak** [comment]
We have some existing code in vg that relies on edge addition being
idempotent, which I'm running into in my attempt to use HashGraph in
more places where vg::VG is used now. The MutableHandleGraph docs say
that it is idempotent, so right now the clearest fix is to fix
HashGraph to match the docs. If we change the docs to match HashGraph,
we don't know what all relied on the old behavior and will need to be
touched.

Plus, to work around the issue, I've been changing code in vg that
creates edges that end up being duplicates to poll with has_edge
first, in order to not upset HashGraph. That's the same O(N) loop, but
in more places. To avoid it, we would need to redesign some algorithms
with not creating duplicate edges in mind.
